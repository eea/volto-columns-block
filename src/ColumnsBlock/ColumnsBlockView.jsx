import React from 'react';
import { useLocation } from 'react-router-dom';
import config from '@plone/volto/registry';
import { Grid } from 'semantic-ui-react';
import { RenderBlocks } from '@plone/volto/components';
import { COLUMNSBLOCK } from '@eeacms/volto-columns-block/constants';
import cx from 'classnames';

import { getColumns } from './utils';
import { getStyle } from '@eeacms/volto-columns-block/Styles';

const getSide = (side, v) => {
  return `${v?.[side] ? `${v[side]}${v.unit ? v.unit : 'px'}` : '0'}`;
};

const getSides = (v) => {
  return `${getSide('top', v)} ${getSide('right', v)} ${getSide(
    'bottom',
    v,
  )} ${getSide('left', v)}`;
};

const ColumnsBlockView = (props) => {
  const location = useLocation();
  const { gridSizes } = config.blocks.blocksConfig[COLUMNSBLOCK];
  const { data = {}, gridSize = 12, gridCols = [] } = props.data;
  const metadata = props.metadata || props.properties;
  const columnList = getColumns(data);
  const customId = props.data?.title
    ?.toLowerCase()
    ?.replace(/[^a-zA-Z-\s]/gi, '')
    ?.trim()
    ?.replace(/\s+/gi, '-');

  return (
    <div className="columns-view" id={customId}>
      <Grid
        columns={gridSize}
        className={
          props.data.reverseWrap ? 'column-grid reverse-wrap' : 'column-grid'
        }
      >
        {columnList.map(([id, column], index) => {
          return (
            <Grid.Column
              key={id}
              {...(gridSizes[gridCols[index]] || gridCols[index])}
              className={cx(
                'column-blocks-wrapper',
                column.settings?.column_class,
              )}
              {...getStyle(column.settings || {})}
            >
              <div
                style={
                  column.settings?.padding
                    ? {
                        padding: getSides(column.settings?.padding),
                      }
                    : {}
                }
              >
                <RenderBlocks
                  {...props}
                  location={location}
                  metadata={metadata}
                  content={column}
                />
              </div>
            </Grid.Column>
          );
        })}
      </Grid>
    </div>
  );
};

export default ColumnsBlockView;
