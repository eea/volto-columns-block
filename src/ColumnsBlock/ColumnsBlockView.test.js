import React from 'react';
// import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import config from '@plone/volto/registry';
import installColumnsBlock from '@eeacms/volto-columns-block';
import { waitFor, render, screen } from '@testing-library/react';

import ColumnsBlockView from './ColumnsBlockView';

const mockStore = configureStore();

// const blockId = '1234';

const blocks = {
  '1234': {
    '@type': 'columnsBlock',
    data: {
      blocks: {
        '9d3b36b7-bc2c-4fd5-80f4-88d104443304': {
          blocks: {
            '421f925b-b89f-4a30-b691-3ee14625101e': {
              '@type': 'text',
              text: 'right',
            },
          },
          blocks_layout: {
            items: ['421f925b-b89f-4a30-b691-3ee14625101e'],
          },
        },
        'b80e32c4-dbd9-481e-a98a-ba42b377e593': {
          blocks: {
            '6cae4500-c1e2-408c-adee-ce87913b072f': {
              '@type': 'text',
              text: 'left marker',
            },
          },
          blocks_layout: {
            items: ['6cae4500-c1e2-408c-adee-ce87913b072f'],
          },
        },
      },
      blocks_layout: {
        items: [
          'b80e32c4-dbd9-481e-a98a-ba42b377e593',
          '9d3b36b7-bc2c-4fd5-80f4-88d104443304',
        ],
      },
    },
    gridCols: ['halfWidth', 'halfWidth'],
    gridSize: 12,
  },
};

const Nop = (props) => <div></div>;
const TextView = ({ data }) => {
  return data.text;
};

test('renders 2 columns', async () => {
  installColumnsBlock(config);
  config.blocks.blocksConfig.text = {
    id: 'text',
    title: 'Text',
    icon: Nop,
    group: 'common',
    view: TextView,
    edit: TextView,
    restricted: false,
    mostUsed: false,
    blockHasOwnFocusManagement: true,
    sidebarTab: 1,
    security: {
      addPermission: [],
      view: [],
    },
  };

  const store = mockStore({
    content: {
      create: {},
      data: {},
      subrequests: {},
    },
    intl: {
      locale: 'en',
      messages: {},
    },
  });

  const { container } = render(
    <Provider store={store}>
      <ColumnsBlockView data={blocks['1234']} metadata={{ blocks }} />
    </Provider>,
  );
  await waitFor(() => screen.getByText('left marker'));
  expect(container).toMatchSnapshot();
});
