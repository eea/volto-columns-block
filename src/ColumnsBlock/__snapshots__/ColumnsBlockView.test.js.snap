// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 2 columns 1`] = `
<div>
  <div
    class="columns-view"
  >
    <div
      class="ui twelve column grid column-grid"
    >
      <div
        class="six wide computer twelve wide mobile six wide tablet column column-blocks-wrapper"
      >
        left marker
      </div>
      <div
        class="six wide computer twelve wide mobile six wide tablet column column-blocks-wrapper"
      >
        right
      </div>
    </div>
  </div>
</div>
`;
